File: bindings/swift/Sources/RNS/RNS.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

import Foundation
import CRNS

public let API_VERSION = "1.5"
public let HASH_LEN = 16

public enum RNSError: Error, Equatable {
    case invalidArgument
    case invalidHandle
    case notFound
    case state
    case io
    case internalError
    case timeout
    case truncated
    case unknown(Int32)

    public var code: Int32 {
        switch self {
        case .invalidArgument: return RNS_ERR_INVALID_ARG
        case .invalidHandle: return RNS_ERR_INVALID_HANDLE
        case .notFound: return RNS_ERR_NOT_FOUND
        case .state: return RNS_ERR_STATE
        case .io: return RNS_ERR_IO
        case .internalError: return RNS_ERR_INTERNAL
        case .timeout: return RNS_ERR_TIMEOUT
        case .truncated: return RNS_ERR_TRUNCATED
        case .unknown(let code): return code
        }
    }

    public static func from(code: Int32) -> RNSError {
        switch code {
        case RNS_ERR_INVALID_ARG: return .invalidArgument
        case RNS_ERR_INVALID_HANDLE: return .invalidHandle
        case RNS_ERR_NOT_FOUND: return .notFound
        case RNS_ERR_STATE: return .state
        case RNS_ERR_IO: return .io
        case RNS_ERR_INTERNAL: return .internalError
        case RNS_ERR_TIMEOUT: return .timeout
        case RNS_ERR_TRUNCATED: return .truncated
        default: return .unknown(code)
        }
    }
}

@discardableResult
public func check(_ code: Int32) throws -> Int32 {
    if code == RNS_OK {
        return code
    }
    throw RNSError.from(code: code)
}

public func version() -> String {
    guard let raw = rns_version() else {
        return ""
    }
    return String(cString: raw)
}

public func lastError() -> String {
    var buf = [CChar](repeating: 0, count: 512)
    var written: Int = 0
    let code = buf.withUnsafeMutableBufferPointer { ptr in
        rns_last_error(ptr.baseAddress, ptr.count, &written)
    }
    if code != RNS_OK || written == 0 {
        return ""
    }
    return String(cString: buf)
}

public func hashToHex(_ data: Data) throws -> String {
    guard data.count == HASH_LEN else {
        throw RNSError.invalidArgument
    }
    return data.map { String(format: "%02x", $0) }.joined()
}

public func hexToHash(_ text: String) throws -> Data {
    guard text.count == 32 else {
        throw RNSError.invalidArgument
    }
    var out = Data(capacity: HASH_LEN)
    var idx = text.startIndex
    for _ in 0..<HASH_LEN {
        let next = text.index(idx, offsetBy: 2)
        guard let byte = UInt8(text[idx..<next], radix: 16) else {
            throw RNSError.invalidArgument
        }
        out.append(byte)
        idx = next
    }
    return out
}

public func hashEq(_ a: Data, _ b: Data) -> Bool {
    a.count == HASH_LEN && b.count == HASH_LEN && a == b
}

File: bindings/java/src/io/quad4/rns/Rns.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

package io.quad4.rns;

import com.sun.jna.ptr.LongByReference;
import io.quad4.rns.ffi.RnsLibrary;

/** Shared constants, version helpers, and error mapping for librns. */
public final class Rns {
    public static final String API_VERSION = "1.5";
    public static final int HASH_LEN = RnsLibrary.HASH_LEN;

    private Rns() {}

    public static String version() {
        String raw = RnsLibrary.INSTANCE.rns_version();
        return raw == null ? "" : raw;
    }

    public static String lastError() {
        byte[] buf = new byte[512];
        LongByReference written = new LongByReference();
        int code = RnsLibrary.INSTANCE.rns_last_error(buf, buf.length, written);
        if (code != RnsLibrary.RNS_OK || written.getValue() == 0) {
            return "";
        }
        return new String(buf, 0, (int) written.getValue());
    }

    public static void check(int code) {
        if (code != RnsLibrary.RNS_OK) {
            throw new RnsException(code);
        }
    }

    public static String hashToHex(byte[] data) {
        if (data == null || data.length != HASH_LEN) {
            throw new RnsException(RnsLibrary.RNS_ERR_INVALID_ARG);
        }
        StringBuilder sb = new StringBuilder(HASH_LEN * 2);
        for (byte b : data) {
            sb.append(String.format("%02x", b & 0xff));
        }
        return sb.toString();
    }

    public static byte[] hexToHash(String text) {
        if (text == null || text.length() != 32) {
            throw new RnsException(RnsLibrary.RNS_ERR_INVALID_ARG);
        }
        byte[] out = new byte[HASH_LEN];
        for (int i = 0; i < HASH_LEN; i++) {
            int hi = Character.digit(text.charAt(i * 2), 16);
            int lo = Character.digit(text.charAt(i * 2 + 1), 16);
            if (hi < 0 || lo < 0) {
                throw new RnsException(RnsLibrary.RNS_ERR_INVALID_ARG);
            }
            out[i] = (byte) ((hi << 4) | lo);
        }
        return out;
    }

    public static boolean hashEq(byte[] a, byte[] b) {
        if (a == null || b == null || a.length != HASH_LEN || b.length != HASH_LEN) {
            return false;
        }
        for (int i = 0; i < HASH_LEN; i++) {
            if (a[i] != b[i]) {
                return false;
            }
        }
        return true;
    }
}
